### PR TITLE
ROX-21630: Create scan config details page

### PR DESCRIPTION
**Pull request by vjwilson**
## Description

This PR implements the read-only view of a single Scan Schedule (`scanConfig` in the API).

Image of the mock from this page:
scanConfig read-only view mock-Compliance 2 0 MVP and beyond

The [API call](https://stagingdb.demo.stackrox.com/main/apidocs-v2#tag/ComplianceScanConfigurationService/operation/ComplianceScanConfigurationService_GetComplianceScanConfiguration) to get this data have, or need, updates, and for MVP the only thing we can get for the profiles are their names.

Notes:

1. `Name` field has to follow RFC 1123, so this PR also adds helper text to the form in the wizard, and a regex and error message for frontend validation.
2. `Description` leave wasn't being sent in POST request when creating Scan Schedules (scanConfigs), so I added that.
3. `Schedule` is using the helper function from the table.
4. `Last Run`: as you can see in the mocks, I added a visual note that this is using the `lastRun` field returned from the API, but I don't think this is correct, is it @dashrews78 ?  I am also using the <Timestamp> component from PatternFly, but this requires 2 CSS utility classes to override some unnecessary opinionated styling that is in that component, to match the other description list items.
5. Profiles widget: as discussed in various project meetings, we do not have data to link to profiles for MVP, and in fact the API response only returns a list of profile names as strings for now, so I am showing them just as a list of strings, instead of the links in the mock.
6. Clusters widget: The only data we are getting for the clusters associated with this scanConfig are:
6.a. `id`: I'm using that as a React key for each row
6.b. `name`: first column of each row
6.c. array of `errors`: not needed in this context?
6.d. There is no data for Type, Provider, Region, and Operator Status. (In a recently-concluded meeting, we said we would remove  Type, Provider, and Region in another context, so we may end up doing that here as well.)
6.e. Of special note, the nested `clusters` array in the scanConfig is not paginated by the API, so I adopted the frontend pagination and sorting for tables where all the data is available up from the PatternFly docs for [Pagination](https://v4-archive.patternfly.org/v4/components/pagination/) and [Table](https://v4-archive.patternfly.org/v4/components/table#composable-sortable--wrapping-headers). 

## Checklist
- [ ] Investigated and inspected CI test results

### Here I tell how I validated my change

Examples
with description
scanConfig with description

without description
scanConfig example with 2 profiles

To test pagination and sorting of the clusters table, I mocked an API response with 41 clusters.

Initial load shows first page
scanConfig-read-view-cluster-pagination

After clicking arrow on first column to reverse its sort
scanConfig-read-view-cluster-pagination-reversed

Advancing to the 2nd page of the current sort
scanConfig-read-view-cluster-pagination-2nd-page

After changing the number displayed per page
scanConfig-read-view-cluster-pagination-changing-per-page

Finally, for the new validation of the name field in the wizard form, here is the error state
frontend validation for scanConfig name

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by dashrews78** [comment]
@vjwilson `lastUpdatedTime` is probably good for a place holder.  We are pulling in another CRD and then we can use https://issues.redhat.com/browse/ROX-21649 to store and return `lastRun`.  So that won't be immediately resolved.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx
Comment (rhmdnd) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx:18]: Will this be wired up to an edit box in a separate commit?

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx
Comment (rhmdnd) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx:113]: This didn't seem to populate for me when I tested this locally. Is there something else I needed to do to get this to render in the UI?

```
 $ ./bin/roxcurl /v2/compliance/scan/configurations | jq '.configurations[].clusterStatus'
[
  {
    "clusterId": "3bbadeca-82ee-4ff5-ae0f-15a89aca3863",
    "errors": [
      ""
    ],
    "clusterName": "remote"
  }
]
[
  {
    "clusterId": "3bbadeca-82ee-4ff5-ae0f-15a89aca3863",
    "errors": [
      ""
    ],
    "clusterName": "remote"
  }
]
```

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx
Comment (rhmdnd) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx:36]: This isn't specific to validation, but I did notice that the UI didn't sort configurations by name in an ascending or descending order.

`compliance-enhanced/cluster-compliance/scan-configs?sortOption[field]=Name&sortOption[direction]=desc` and `compliance-enhanced/cluster-compliance/scan-configs?sortOption[field]=Name&sortOption[direction]=asc` present the same order with two separate scan configs.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx:18]: Yes, this PR is just to get the read-only view working.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx:113]: I was expecting some kind "status" value returned from the API. Is it simply that an empty `errors` array is healthy, and a non-empty `errors` array is unhealthy?

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx:36]: I'll check that, but if it's not a regression in this PR, I'll just file it a bug, to keep the changes on topic.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx
Comment (rhmdnd) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx:113]: That seems reasonable, but I'll let @dashrews78 weigh-in on if there is another part of the ACS API dedicated to serving up the cluster health.

There is the integrations API:

```
$ ./bin/roxcurl /v2/compliance/integrations | jq .
{
  "integrations": [
    {
      "id": "8469f1f9-af2a-4ebf-bb96-28c0202f8639",
      "version": "v1.4.0",
      "clusterId": "3bbadeca-82ee-4ff5-ae0f-15a89aca3863",
      "clusterName": "remote",
      "namespace": "openshift-compliance",
      "statusErrors": []
    }
  ]
}
```

But that has a similar pattern where it looks like we infer healthy based on the absence of anything in the `statusErrors` list.

Comment (dashrews78) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx:113]: Yeah I the presence or not of the `errors` is how status can be inferred.  I suppose we could add that to the API if you really need us to.

Comment (rhmdnd) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ViewScanConfigDetail.tsx:113]: Tested this with the latest commit and it worked for me locally. Thanks!

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigClustersTable.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigClustersTable.tsx:78]: Minor, but what do you think about the following to remove the need for `as` here:
```suggestion
        if (typeof aValue === 'number' && typeof bValue === 'number') {
            // Numeric sort
            if (activeSortDirection === 'asc') {
                return aValue - bValue;
            }
            return bValue - aValue;
        }
        if (typeof aValue === 'string' && typeof bValue === 'string') {
            // String sort
            if (activeSortDirection === 'asc') {
                return aValue.localeCompare(bValue);
            }
            return bValue.localeCompare(aValue);
        }
```
Given the logic it shouldn't ever happen, but if somehow `aValue` was a `string` and `bValue` was a `number`, the second case here would throw a runtime error.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigClustersTable.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigClustersTable.tsx:57]: Too bad there isn't an eslint equivalent to `@ts-expect-error` that would error out here once the API is implemented, as an alternative to the `as` type assertion.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx:41]: Does this satisfy the constraint "end with an alphanumeric character"? From what I can tell that isn't the case:
```ts
/(^((?!-)(?!\.)[a-z0-9-.]{0,253}$))/.test("scan.");
// true
```

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx:11]:
```suggestion
const EditScanConfigButton = (id: string) => {
```
Nit to remove implicit `any`

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigParameters.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigParameters.tsx:1]:
```suggestion
/* eslint-disable no-nested-ternary */
```
This looks unused too.

(Honestly there are enough of these comments that I'd say we can disable the rule entirely.)

**Review by dvail — APPROVED** [review]
Approving since most of my comments are nits that I think will go away naturally once this is fully implemented.

The only thing we should double check before merging that I can see is that the regex used matches what is expected.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/ScanConfigDetailPage.tsx:11]: Agree, but this is actually an artifact that isn't used in this file anymore; I moved it to `ViewScanConfigDetail.tsx`. I'll delete it here.

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/Wizard/useFormikScanConfig.tsx:41]: Doh! I was so concerned with the negative lookahead at the front, I forgot about the back. 

In a effort not to complicate it any more, I think I will instead go with something simpler, like this
`^[a-z0-9][a-z0-9.-]{0,251}[a-z0-9]$`

Address the ending character limitation, although it does require at least 2 characters for the name.  (probably a good practice anyway)

Screenshot 2024-01-19 at 4 53 06 PM

File: ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigProfiles.tsx
Comment (vjwilson) [reply ui/apps/platform/src/Containers/ComplianceEnhanced/ClusterCompliance/ScanConfigs/components/ScanConfigProfiles.tsx:11]: I started out with the Card layouts in the parent component, but then ran into some convoluted data handling for the nested Clusters table. So, I decided it was better for all 3 components to have parallel structure.

**Pull request merged**
